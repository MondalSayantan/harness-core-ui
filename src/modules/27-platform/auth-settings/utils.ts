/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { get } from 'lodash-es'
import type { CreateIpAllowlistConfigProps, UpdateIpAllowlistConfigProps } from '@harnessio/react-ng-manager-client'

import type { IIPAllowlistForm } from '@auth-settings/interfaces/IPAllowlistInterface'
import { SourceType } from '@auth-settings/interfaces/IPAllowlistInterface'
import { CurrencyType } from '@common/constants/SubscriptionTypes'

export function getAmountInCurrency(currency: CurrencyType, amount: number): string {
  switch (currency) {
    case CurrencyType.USD: {
      return amount.toLocaleString('en-US', {
        style: 'currency',
        currency: 'USD'
      })
    }
  }

  return ''
}

// dividing the actual amount by 100 as amount numbers are in cents
export function getDollarAmount(amount?: number, isYearly?: boolean): number {
  if (isYearly) {
    return (amount || 0) / 1200
  }
  return (amount || 0) / 100
}

export const getErrorMessage = (error: any): string =>
  get(error, 'data.error', get(error, 'data.message', error?.message))

export const buildCreateIPAllowlistPayload = (data: IIPAllowlistForm): CreateIpAllowlistConfigProps => {
  const allowSourceType = []
  if (data.allowSourceTypeUI) {
    allowSourceType.push(SourceType.UI)
  }
  if (data.allowSourceTypeAPI) {
    allowSourceType.push(SourceType.API)
  }
  return {
    body: {
      ip_allowlist_config: {
        allowed_source_type: allowSourceType,
        description: data?.description,
        enabled: false, // always false while creating
        identifier: data.identifier,
        ip_address: data.ipAddress,
        name: data.name,
        tags: data?.tags
      }
    }
  }
}

export const buildUpdateIPAllowlistPayload = (
  data: IIPAllowlistForm,
  overrideFields?: any
): UpdateIpAllowlistConfigProps => {
  const allowSourceType = []
  if (data.allowSourceTypeUI) {
    allowSourceType.push(SourceType.UI)
  }
  if (data.allowSourceTypeAPI) {
    allowSourceType.push(SourceType.API)
  }
  return {
    'ip-config-identifier': data.identifier,
    body: {
      ip_allowlist_config: {
        allowed_source_type: allowSourceType,
        description: data?.description,
        enabled: data?.enabled,
        identifier: data.identifier,
        ip_address: data.ipAddress,
        name: data.name,
        tags: data?.tags,
        ...overrideFields
      }
    }
  }
}

/**
 * Converts minutes to higher dimensions of time such as months, days and hours
 *
 * Examples:
 * formatMinutesToHigherDimensions(150);    // Output: "2 hours 30 minutes"
 * formatMinutesToHigherDimensions(4320);   // Output: "3 days"
 * formatMinutesToHigherDimensions(45000);  // Output: "1 month 1 day 6 hours"
 * @param minutes
 * @returns a string in higher dimensions
 */
export const formatMinutesToHigherDimensions = (minutes: number | undefined): string => {
  if (minutes === undefined || minutes < 0) {
    return 'Invalid input'
  }

  const MINUTES_IN_AN_HOUR = 60
  const HOURS_IN_A_DAY = 24
  const DAYS_IN_A_MONTH = 30

  const result = []

  if (minutes >= MINUTES_IN_AN_HOUR * HOURS_IN_A_DAY * DAYS_IN_A_MONTH) {
    const months = Math.floor(minutes / (MINUTES_IN_AN_HOUR * HOURS_IN_A_DAY * DAYS_IN_A_MONTH))
    result.push(`${months} month${months > 1 ? 's' : ''}`)
    minutes %= MINUTES_IN_AN_HOUR * HOURS_IN_A_DAY * DAYS_IN_A_MONTH
  }

  if (minutes >= MINUTES_IN_AN_HOUR * HOURS_IN_A_DAY) {
    const days = Math.floor(minutes / (MINUTES_IN_AN_HOUR * HOURS_IN_A_DAY))
    result.push(`${days} day${days > 1 ? 's' : ''}`)
    minutes %= MINUTES_IN_AN_HOUR * HOURS_IN_A_DAY
  }

  if (minutes >= MINUTES_IN_AN_HOUR) {
    const hours = Math.floor(minutes / MINUTES_IN_AN_HOUR)
    result.push(`${hours} hour${hours > 1 ? 's' : ''}`)
    minutes %= MINUTES_IN_AN_HOUR
  }

  if (minutes > 0) {
    result.push(`${minutes} minute${minutes > 1 ? 's' : ''}`)
  }

  return result.join(' ')
}
