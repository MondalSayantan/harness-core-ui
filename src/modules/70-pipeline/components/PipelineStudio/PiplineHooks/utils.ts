/*
 * Copyright 2023 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { defaultTo, flatMapDeep, identity, map, sortBy, sortedUniq } from 'lodash-es'
import { VariableMergeServiceResponse, VariableResponseMapValue } from 'services/pipeline-ng'

export interface Params {
  metadataMap: Record<string, VariableResponseMapValue>
  localStageKeys: string[]
}

export interface AllExpressionsObj {
  expressionsList: string[]
  outputExpressions: string[]
  extraExpressions: string[]
  extraOutputExpressions: string[]
}

/**
 * Traverse over stage and find out all local fqn
 */
export function traverseStageObject(
  jsonObj: Record<string, any>,
  metadataMap: Required<VariableMergeServiceResponse>['metadataMap']
): string[] {
  const keys: string[] = []
  if (jsonObj !== null && typeof jsonObj === 'object') {
    Object.entries(jsonObj).forEach(([_key, value]) => {
      keys.push(...traverseStageObject(value, metadataMap))
    })
  } else if (metadataMap[jsonObj]) {
    keys.push(jsonObj)
  }
  return keys
}

export function pickExpressionsFromMetadataMap(
  metadataMap: Record<string, VariableResponseMapValue>,
  localStageKeys: string[],
  key: 'yamlProperties' | 'yamlOutputProperties'
): string[] {
  return sortedUniq(
    sortBy(
      map(metadataMap, (item, index) =>
        index && localStageKeys.indexOf(index) > -1
          ? defaultTo(item[key]?.localName, '')
          : defaultTo(item[key]?.fqn, '')
      ).filter(p => p),
      identity
    )
  )
}

export function pickExtraExpressionsFromMetadataMap(
  metadataMap: Record<string, VariableResponseMapValue>,
  localStageKeys: string[],
  key: 'properties' | 'outputproperties'
): string[] {
  return sortedUniq(
    sortBy(
      flatMapDeep(metadataMap, (item, index) => {
        const properties = defaultTo(item.yamlExtraProperties?.[key], [])

        return properties.map((p: Record<string, string>) =>
          index && localStageKeys.indexOf(index) > -1 ? p.localName : p.fqn
        )
      }).filter(p => p),
      identity
    )
  )
}

export const getAllExpressionsFromMetadataMap = ({ metadataMap, localStageKeys }: Params): AllExpressionsObj => ({
  expressionsList: pickExpressionsFromMetadataMap(metadataMap, localStageKeys, 'yamlProperties'),
  outputExpressions: pickExpressionsFromMetadataMap(metadataMap, localStageKeys, 'yamlOutputProperties'),
  extraExpressions: pickExtraExpressionsFromMetadataMap(metadataMap, localStageKeys, 'properties'),
  extraOutputExpressions: pickExtraExpressionsFromMetadataMap(metadataMap, localStageKeys, 'outputproperties')
})
