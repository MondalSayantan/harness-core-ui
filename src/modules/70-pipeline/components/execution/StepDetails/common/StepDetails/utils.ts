/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { ResponseApprovalInstanceResponse } from 'services/pipeline-ng'

export function encodeURIWithReservedChars(uri: string): string {
  return encodeURIComponent(uri).replace(/[;,/?:@&=+$#]/g, function (c) {
    return '%' + c.charCodeAt(0).toString(16)
  })
}

export interface ExecutionMetadataType {
  approvalInstanceId: string
  mock?: {
    data?: ResponseApprovalInstanceResponse
    loading?: boolean
  }
}
