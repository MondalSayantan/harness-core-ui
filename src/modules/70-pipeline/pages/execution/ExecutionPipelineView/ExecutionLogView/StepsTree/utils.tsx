import { defaultTo } from 'lodash-es'
import { ExecutionPipelineNode } from '@pipeline/components/ExecutionStageDiagram/ExecutionPipelineModel'
import { ExecutionNode, InterruptEffectDTO } from 'services/pipeline-ng'

export function getRetryInterrupts(step: ExecutionPipelineNode<ExecutionNode>): InterruptEffectDTO[] {
  return defaultTo(step?.item?.data?.interruptHistories, defaultTo(step?.group?.data?.interruptHistories, [])).filter(
    row => row.interruptType === 'RETRY'
  )
}
