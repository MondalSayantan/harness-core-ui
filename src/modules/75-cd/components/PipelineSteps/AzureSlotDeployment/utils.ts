import { RUNTIME_INPUT_VALUE, MultiTypeInputType, getMultiTypeFromValue } from '@harness/uicore'
import type { StageElementWrapper } from '@pipeline/utils/pipelineTypes'
import type { StageElementConfig, StageElementWrapperConfig, PipelineInfoConfig } from 'services/pipeline-ng'
import type { InfraStructureDefinitionYaml, DeploymentStageConfig, EnvironmentYamlV2 } from 'services/cd-ng'

interface EnvironmentYamlV2Runtime extends Omit<EnvironmentYamlV2, 'infrastructureDefinitions'> {
  infrastructureDefinitions?: InfraStructureDefinitionYaml[] | string
}

interface DeploymentStageConfigAzureWebName extends Omit<DeploymentStageConfig, 'environment'> {
  environment: EnvironmentYamlV2Runtime
}

export interface DeploymentStageElementConfigAzureWebName extends Omit<StageElementConfig, 'spec'> {
  spec?: DeploymentStageConfigAzureWebName
}
export type SelectedStageType = StageElementWrapper<DeploymentStageElementConfigAzureWebName>

export function getStage(stageId: string, pipeline?: PipelineInfoConfig): StageElementWrapperConfig | undefined {
  if (pipeline?.stages) {
    let responseStage: StageElementWrapperConfig | undefined = undefined
    pipeline.stages.forEach(item => {
      if (item.stage && item.stage.identifier === stageId) {
        responseStage = item
      } else if (item.parallel) {
        return item.parallel.forEach(node => {
          if (node.stage?.identifier === stageId) {
            responseStage = node
          }
        })
      }
    })
    return responseStage
  }
  return
}

export const isRuntimeEnvId = (selectedStage: SelectedStageType): boolean => {
  const envId = selectedStage?.stage?.spec?.environment?.environmentRef
  return !!(envId && envId === RUNTIME_INPUT_VALUE)
}

export const isRuntimeInfraId = (selectedStage: SelectedStageType): boolean => {
  const infraId = selectedStage?.stage?.spec?.environment?.infrastructureDefinitions
  return infraId === RUNTIME_INPUT_VALUE
}

export const isMultiEnv = (selectedStage: SelectedStageType): boolean => {
  const isEnvs = !!selectedStage?.stage?.spec?.environments || !!selectedStage?.stage?.spec?.environmentGroup
  return !!isEnvs
}

export const getEnvId = (selectedStage: SelectedStageType): string => {
  if (isMultiEnv(selectedStage || isRuntimeEnvId(selectedStage))) {
    return ''
  }

  return (
    selectedStage?.stage?.spec?.environment?.environmentRef ||
    selectedStage?.stage?.template?.templateInputs?.spec?.environment?.environmentRef ||
    ''
  )
}

export const getInfraId = (selectedStage: SelectedStageType, infraSourceField?: string): string => {
  const envId = getEnvId(selectedStage)
  if (!envId || isMultiEnv(selectedStage)) {
    return ''
  }

  if (!isRuntimeEnvId(selectedStage) && !isRuntimeInfraId(selectedStage)) {
    const [infra] = selectedStage?.stage?.spec?.environment?.infrastructureDefinitions as InfraStructureDefinitionYaml[]
    if (infraSourceField) {
      return infra?.inputs?.spec?.[infraSourceField]
    }
    return infra.identifier
  }
  return ''
}

/* eslint-disable */
export const getEnvIdRuntime = (stageId: string, values: any): string => {
  const stage = getStage(stageId, values) as SelectedStageType
  const envId =
    stage?.stage?.spec?.environment?.environmentRef ||
    stage?.stage?.template?.templateInputs?.spec?.environment?.environmentRef
  return envId || ''
}
/* eslint-disable */
export const getInfraIdRuntime = (stageId: string, values: any, infraSourceField?: string): string => {
  const stage = getStage(stageId, values) as SelectedStageType
  const isTemplateStageInfra = stage?.stage?.template?.templateInputs?.spec?.environment?.infrastructureDefinitions?.[0]
  if (isTemplateStageInfra) {
    const [{ inputs, identifier }] = stage?.stage?.template?.templateInputs?.spec?.environment
      ?.infrastructureDefinitions as InfraStructureDefinitionYaml[]
    if (infraSourceField && infraSourceField === 'infrasctructure') {
      return identifier
    }
    if (infraSourceField && inputs?.spec) {
      return inputs?.spec?.[infraSourceField]
    }
  }

  if (stage?.stage?.spec?.environment?.infrastructureDefinitions?.[0]) {
    const [{ identifier, inputs }] = stage?.stage?.spec?.environment
      ?.infrastructureDefinitions as InfraStructureDefinitionYaml[]
    if (infraSourceField && inputs?.spec && infraSourceField !== 'infrasctructure') {
      return inputs?.spec?.[infraSourceField]
    }

    return identifier
  }
  return ''
}

export const getAllowableTypes = (selectedStage: SelectedStageType): MultiTypeInputType[] => {
  if (isMultiEnv(selectedStage)) {
    return [MultiTypeInputType.EXPRESSION]
  }

  return [MultiTypeInputType.FIXED, MultiTypeInputType.RUNTIME, MultiTypeInputType.EXPRESSION]
}

export const infrastructurePath = 'template.templateInputs.spec.environment.infrastructureDefinitions[0]'
export const resourceGroupPath = `${infrastructurePath}.inputs.spec.resourceGroup`
export const connectorPath = `${infrastructurePath}.inputs.spec.connectorRef`
export const subscriptionPath = `${infrastructurePath}.inputs.spec.subscriptionId`

export const getInfraParamsFixedValue = (values: string[]) => {
  let fixedValue = ''
  values?.forEach((value: string) => {
    if (value && getMultiTypeFromValue(value) === MultiTypeInputType.FIXED) {
      fixedValue = value
    }
  })
  return fixedValue
}
