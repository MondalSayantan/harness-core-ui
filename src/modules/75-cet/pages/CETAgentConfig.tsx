/*
 * Copyright 2023 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { ErrorTracking } from '@cet/ErrorTrackingApp'
import ChildAppMounter from 'microfrontends/ChildAppMounter'
import { useFeatureFlags } from '@common/hooks/useFeatureFlag'
import CardWithOuterTitle from '@common/components/CardWithOuterTitle/CardWithOuterTitle'
import { CETAgentConfigProps } from '@cet/ErrorTracking.types'

export const CETAgentConfig: React.FC<CETAgentConfigProps> = props => {
  const { CET_PLATFORM_MONITORED_SERVICE } = useFeatureFlags()

  if (CET_PLATFORM_MONITORED_SERVICE) {
    return (
      <CardWithOuterTitle>
        <ChildAppMounter
          data-testid={'error-tracking-child-mounter'}
          ChildApp={ErrorTracking}
          componentLocation={'agent-config'}
          monitoredService={{
            serviceRef: props.serviceRef,
            environmentRef: props.environmentRef
          }}
        />
      </CardWithOuterTitle>
    )
  } else {
    return <></>
  }
}
