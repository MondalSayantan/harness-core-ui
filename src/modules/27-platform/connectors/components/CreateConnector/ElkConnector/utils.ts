/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import type { FormData } from '@platform/connectors/interfaces/ConnectorInterface'
import { AuthTypes } from '@platform/connectors/pages/connectors/utils/ConnectorHelper'
import { UseStringsReturn } from 'framework/strings'
import type { ConnectorConfigDTO } from 'services/cd-ng'
import { ElkConnectorAuthFieldsProps } from './CreateElkConnector.types'
import { ELKConnectorFields, ElkAuthType } from './CreateElkConnector.constants'

type InitializeElkConnectorArgs = {
  prevStepData?: ConnectorConfigDTO
  accountId: string
  projectIdentifier?: string
  orgIdentifier?: string
}

export function initializeElkConnector({
  prevStepData,
  accountId,
  projectIdentifier,
  orgIdentifier
}: InitializeElkConnectorArgs): FormData {
  const defaultObj = {
    url: '',
    accountId,
    username: '',
    authType: AuthTypes.USER_PASSWORD,
    password: undefined,
    projectIdentifier,
    orgIdentifier
  }

  if (!prevStepData) {
    return defaultObj
  }
  const { spec, ...prevData } = prevStepData
  return {
    ...defaultObj,
    url: prevData?.url || spec?.controllerUrl || spec?.url || '',
    password: prevData?.password || spec?.passwordRef,
    clientSecretRef: prevData?.clientSecretRef || spec?.clientSecretRef,
    authType: prevData?.authType || spec?.authType || AuthTypes.USER_PASSWORD,
    username: prevData?.username || spec?.username || '',
    clientId: prevData?.clientId || spec?.clientId,
    apiKeyId: prevData?.apiKeyId || spec?.apiKeyId,
    apiKeyRef: prevData?.apiKeyRef || spec?.apiKeyRef
  }
}

interface GetElkAuthTypeReturn {
  label: string
  value: string
}

export const getELKAuthType = (getString: UseStringsReturn['getString']): GetElkAuthTypeReturn[] => [
  {
    label: getString('usernamePassword'),
    value: AuthTypes.USER_PASSWORD
  },
  {
    label: getString('common.apikey'),
    value: AuthTypes.API_CLIENT_TOKEN
  },
  {
    label: getString('platform.connectors.bearerToken'),
    value: AuthTypes.BEARER_TOKEN
  },
  {
    label: getString('platform.connectors.elk.noAuthentication'),
    value: ElkAuthType.NONE
  }
]

export const getAuthFields = (
  getString: UseStringsReturn['getString'],
  authType?: string
): ElkConnectorAuthFieldsProps[] | null => {
  if (!authType) {
    return null
  }

  switch (authType) {
    case AuthTypes.USER_PASSWORD:
      return [
        { name: ELKConnectorFields.USERNAME, label: getString('username'), key: ELKConnectorFields.USERNAME },
        { name: ELKConnectorFields.PASSWORD, label: getString('password'), key: ELKConnectorFields.PASSWORD }
      ]

    case AuthTypes.API_CLIENT_TOKEN:
      return [
        {
          name: ELKConnectorFields.API_KEY_ID,
          label: getString('platform.connectors.elk.apiId'),
          key: ELKConnectorFields.API_KEY_ID
        },
        {
          name: ELKConnectorFields.API_KEY_REF,
          label: getString('common.apikey'),
          key: ELKConnectorFields.API_KEY_REF
        }
      ]

    case AuthTypes.BEARER_TOKEN:
      return [
        {
          name: ELKConnectorFields.API_KEY_REF,
          label: getString('platform.connectors.bearerToken'),
          key: ELKConnectorFields.API_KEY_REF
        }
      ]

    default:
      return null
  }
}
