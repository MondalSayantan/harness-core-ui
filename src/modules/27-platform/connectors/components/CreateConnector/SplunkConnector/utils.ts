/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import type { FormData } from '@platform/connectors/interfaces/ConnectorInterface'
import { AuthTypes } from '@platform/connectors/pages/connectors/utils/ConnectorHelper'
import type { ConnectorConfigDTO } from 'services/cd-ng'

type InitializeSplunkConnectorArgs = {
  prevStepData?: ConnectorConfigDTO
  accountId: string
  projectIdentifier?: string
  orgIdentifier?: string
}

export function initializeSplunkConnector({
  prevStepData,
  projectIdentifier,
  accountId,
  orgIdentifier
}: InitializeSplunkConnectorArgs): FormData {
  const defaultObj = {
    url: '',
    authType: AuthTypes.USER_PASSWORD,
    username: '',
    passwordRef: undefined,
    tokenRef: undefined,
    accountId: accountId,
    projectIdentifier,
    orgIdentifier
  }

  if (!prevStepData) {
    return defaultObj
  }

  const { spec, ...prevData } = prevStepData
  return {
    ...defaultObj,
    url: prevData?.url || spec?.splunkUrl || '',
    username: prevData?.username || spec?.username,
    passwordRef: prevData?.passwordRef || spec?.passwordRef,
    authType: prevData?.authType || spec?.type || defaultObj.authType,
    tokenRef: prevData?.tokenRef || spec?.tokenRef
  }
}
