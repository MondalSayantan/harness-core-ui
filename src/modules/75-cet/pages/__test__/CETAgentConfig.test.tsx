/*
 * Copyright 2023 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { render } from '@testing-library/react'
import { TestWrapper, TestWrapperProps } from '@common/utils/testUtils'
import * as hooks from '@common/hooks/useFeatureFlag'
import routes from '@common/RouteDefinitions'
import { accountPathProps, orgPathProps, projectPathProps } from '@common/utils/routeUtils'
import { CETAgentConfig } from '../CETAgentConfig'

// Mock the ChildAppMounter component
jest.mock('microfrontends/ChildAppMounter', () => () => {
  return <div data-testid="error-tracking-child-mounter">mounted</div>
})

describe('CETAgentConfig', () => {
  // Mock the useFeatureFlags hook to return a specific value
  jest.spyOn(hooks, 'useFeatureFlags').mockReturnValue({
    CET_PLATFORM_MONITORED_SERVICE: true // Set to true for this test
  })

  const testWrapperProps: TestWrapperProps = {
    path: routes.toErrorTracking({ ...accountPathProps, ...orgPathProps, ...projectPathProps }),
    pathParams: {
      accountId: 'test_accountId',
      projectIdentifier: 'test_project',
      orgIdentifier: 'test_org'
    }
  }

  test('Verify if Error Tracking View is rendered when required params are defined', async () => {
    const props = {
      serviceRef: 'serviceRef',
      environmentRef: 'environmentRef'
    }

    const { getByTestId } = render(
      <TestWrapper {...testWrapperProps}>
        <CETAgentConfig {...props} />
      </TestWrapper>
    )

    const childAppMounter = getByTestId('error-tracking-child-mounter')
    expect(childAppMounter).toBeTruthy()
  })

  test('Verify if nothing is rendered when CET_PLATFORM_MONITORED_SERVICE is false', async () => {
    jest.spyOn(hooks, 'useFeatureFlags').mockReturnValue({
      CET_PLATFORM_MONITORED_SERVICE: false
    })

    const { container } = render(
      <TestWrapper {...testWrapperProps}>
        <CETAgentConfig />
      </TestWrapper>
    )

    expect(container.firstChild).toBeNull()
  })

  test('Snapshot test for CETAgentConfig component', () => {
    jest.spyOn(hooks, 'useFeatureFlags').mockReturnValue({
      CET_PLATFORM_MONITORED_SERVICE: false
    })

    const { asFragment } = render(
      <TestWrapper {...testWrapperProps}>
        <CETAgentConfig />
      </TestWrapper>
    )

    // Take a snapshot of the rendered component
    expect(asFragment()).toMatchSnapshot()
  })
})
