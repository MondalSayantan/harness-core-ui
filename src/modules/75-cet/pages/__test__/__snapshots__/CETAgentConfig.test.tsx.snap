// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CETAgentConfig Snapshot test for CETAgentConfig component 1`] = `<DocumentFragment />`;
